### PR TITLE
Session construct WIP

### PR DESCRIPTION
**Pull request by vjardin**
For comments

### CODE REVIEW

File: pkcs11module.c [hunk 1]
```diff
@@ -149,6 +152,20 @@ PHP_METHOD(Module, __construct) {
...
+
+    if (!objval->initialised)
+        return;
+
+    rv = objval->functionList->C_Finalize(NULL);
+    if ((rv != CKR_OK) && (rv != CKR_CRYPTOKI_NOT_INITIALIZED))
+        pkcs11_error(rv, "could not be C_Finalize()'d");
+
+    objval->initialised = false;
+}
```
Comment (gamringer) [inline pkcs11module.c:168]: This is redundant since C_Finalize is already being called inpkcs11_shutdown()

File: pkcs11session.c [hunk 2]
```diff
@@ -79,23 +100,141 @@ ZEND_BEGIN_ARG_INFO_EX(arginfo_destroyObject, 0, 0, 1)
...
     pkcs11_session_object *objval = Z_PKCS11_SESSION_P(ZEND_THIS);
+
+    CK_RV rv;
+
+    ZEND_PARSE_PARAMETERS_NONE();
+
+    rv = objval->pkcs11->functionList->C_CloseSession(objval->session);
+    if (rv != CKR_OK)
+        pkcs11_error(rv, "could not be C_CloseSession()'d");
+}
```
Comment (gamringer) [inline pkcs11session.c:156]: C_CloseSession is already called in pkcs11_session_shutdown(). This is redundant.

**Comment by gamringer** [comment]
I don't think that we are on the same page regarding implementing the C_* function to keep the original API with as much fidelity as possible. The was that I was seeing it, all C_* functions would be called as methds of the Module class.the Session object would be passed as first argument to all C_* function needing it, just like the original API. If the only thing that we want from the original API is the CK_RV in case of errors, then I think a much better approach would be to have the CK_RV set as the exception code in the existing methods. We would have to start throwing a new Pkcs11Exception class for this instead of relying on the default exception (which we should do regardless of the topic at hand)

**Comment by vjardin** [comment]
I agree that we need to be close to the OASIS semantics in order to apply easily any OASIS recommendation like the ones from Thales/Safenet https://thalesdocs.com/gphsm/ptk/5.5/docs/Content/PTK-C_Program/PKCS_11_Cmd_Ref/Slot_Token_Mgmt/C_GetSlotInfo.htm or any other ones.

The C_* samantic is a must have for a long term supportability.

Nevertheless, starting to map, port and play with the php-pkcs11 I am able to easily use it thanks to the same C_* methods. However, I was having some issues to manage a close of the sessions in case of any PHP errors before close() could be called. That's why I started to try getting the automatic PHP garbage collectors that would handle the calls to close() using the destructors that are called 'automatically'.

If we keep all the C_* methods into the Module class, how to handle property errors that should require a proper close()  ?

I agree that there is a specific topic for CK_RV: returned values vs a Pkcs11 Exception code but I feel it should be postponed after some more trials. I am wondering, like for soap classes, if we should have both modes: returned error and throwing exception. I think I could support it using a flag with the constructor of the module.

File: pkcs11module.c (hunk 1 above)
Comment (vjardin) [reply pkcs11module.c:168]: OK, I think got the point of pkcs11_shutdown(): first time I thought it was called only when  php-pkcs11 gets shutdown by PHP. I see it registers for the destructor too thanks to  DEFINE_MAGIC_FUNCS() / DECLARE_MAGIC_FUNCS().

 I missed it.

Then, what's about adding warning case pkcs11_error(rv, "could not be C_Finalize()'d"); within pkcs11_shutdown() ?

File: pkcs11session.c (hunk 2 above)
Comment (vjardin) [reply pkcs11session.c:156]: you are right: Same as the previous comment, I missed the magics from DEFINE_MAGIC_FUNCS() / DECLARE_MAGIC_FUNCS().

Again, what's about adding if (rv != CKR_OK) pkcs11_error(rv, "could not be C_CloseSession()'d"); ?

**Comment by vjardin** [comment]
Back to my previous comment: I was missing your MAGIC macros. Please, wait till I'll perform more tests in order to get the full picture. Let's close this serie. I'll send another one for our thoughts.

**Pull request closed without merging**